yui/apps/weather/weather.py

```diff
@@ -0,0 +1,427 @@
+from ...event import Message
+from ...utils import json
+from ...utils.datetime import now
+
+box.assert_config_required("GOOGLE_API_KEY", str)
+box.assert_config_required("OPENWEATHER_API_KEY", str)
+
+
+class WeatherResponseError(RuntimeError):
+    pass
+
+
+EXCEPTIONS = (
+    client_exceptions.ClientPayloadError,  # Bad HTTP Response
+    ValueError,  # JSON Error
+    client_exceptions.ClientConnectorCertificateError,  # TLS expired
+    WeatherResponseError,  # Bad HTTP Response
+)
+
+
+LABELS = {
+    "pm25": "PM2.5",
+    "pm10": "PM10",
+    "o3": "오존",
+    "no": "일산화 질소",
+    "no2": "이산화 질소",
+    "so2": "이산화 황",
+    "co": "일산화 탄소",
+    "nh3": "암모니아",
+}
+
+
+@define
+class WeatherRecord:
+    # 위치에 대한 정보 (영어)
+    location: str
+
+    # 현재, 최고/최소 온도, 습도 및 기압, 가시거리
+    current_temp: float
+    min_temp: float
+    max_temp: float
+    feel_temp: float
+    humidity: int
+    pressure: float
+    visibility: int
+
+    # 바람의 세기와 각도
+    wind_speed: float
+    wind_degree: int
+
+    # 구름 정도
+    cloudiness: int
+
+    # 강우 유무
+    rain: bool
+    rain_1h: Optional[float]
+    rain_3h: Optional[float]
+
+    # 강설 유무
+    snow: bool
+    snow_1h: Optional[float]
+    snow_3h: Optional[float]
+
+    # 날씨 설명 (영어)
+    status: str
+    description: str
+
+
+@define
+class AirPollutionRecord:
+
+    aqi: int  # 1~5까지의 AQI Index
+    co: Optional[float] = None  # 일산화 탄소 (Carbon Monoxide)
+    no: Optional[float] = None  # 일산화 질소
+    no2: Optional[float] = None  # 이산화 질소 (Nitrogen Dioxide)
+    o3: Optional[float] = None  # 오존(Ozone)
+    so2: Optional[float] = None  # 이산화 황 (Sulphur Dioxide)
+    pm25: Optional[float] = None  # PM2.5
+    pm10: Optional[float] = None  # PM10
+    nh3: Optional[float] = None  # 암모니아
+
+
+async def get_geometric_info_by_address(
+    address: str,
+    api_key: str,
+) -> tuple[str, float, float]:
+    url = "https://maps.googleapis.com/maps/api/geocode/json?" + urlencode(
+        {"region": "kr", "address": address, "key": api_key}
+    )
+    async with aiohttp.ClientSession(
+        headers={"Accept-Language": "ko-KR"}
+    ) as session:
+        async with session.get(url) as resp:
+            if resp.status != 200:
+                raise WeatherResponseError(f"Bad HTTP Response: {resp.status}")
+
+            data = await resp.json(loads=json.loads)
+
+    full_address = data["results"][0]["formatted_address"]
+    lat = data["results"][0]["geometry"]["location"]["lat"]
+    lng = data["results"][0]["geometry"]["location"]["lng"]
+
+    return full_address, lat, lng
+
+
+async def get_weather_by_coordinate(
+    lat: float,
+    lng: float,
+    api_key: str,
+) -> WeatherRecord:
+    url = "https://api.openweathermap.org/data/2.5/weather?" + urlencode(
+        {
+            "lat": lat,
+            "lon": lng,
+            "appid": api_key,
+            "units": "metric",
+        }
+    )
+
+    async with aiohttp.ClientSession() as session:
+        async with session.get(url) as resp:
```

Comment:
저도 최근에서야 회사 코드리뷰 하다가 접목할 생각을 해본거지만...

```suggestion
    url = "https://api.openweathermap.org/data/2.5/weather"
    params = {
        "lat": lat,
        "lon": lng,
        "appid": api_key,
        "units": "metric",
    }
 
    async with aiohttp.ClientSession() as session:
        async with session.get(url, params=params) as resp:
```

그런데 query param의 `lon` 은 오타가 아닌가요?

Reply:
아니에요. 저기는 파라미터를 `lon`으로 받더라구요. URL이랑 쿼리스트링을 분리하는건 대찬성이에요. 제가 httpx만 써서 잘 몰랐어요.